### DISCUSSION TITLE
请教一个gRPC套cdn后fullcone的设置问题

### DISCUSSION THREAD

**Discussion by wtfr-dot** [XTLS/Xray-core#3358] [Q&A]
看了一些说明和讨论后，我明白现在xray的fullcone是支持过cdn的，但没有一个明确的设置方法，我看了一下passwall的设置在采用gRPC的底层传输时有mux选项，但没有xudp的设置，我目前的设置是开启mux，concurrency为-1，我的理解是tcp不使用mux，udp走mux，xudp是使用了mux的，那这样设置就是xudp吗？还是只是单纯的UoT？是否能fullcone？我测试出来的结果是nat3
另外我看V2rayN的设置里，不开启mux的情况下默认也有一个"packet_encoding": "xudp"参数，这个参数的作用是开启xudp的支持吗？为什么又跟mux没有关系？
我又在#252里看见特性描述里有这样一点：客户端、服务端均为 Xray-core v1.3.0+ 时，VLESS、VMess、Mux 默认均为 FullCone，这是否是说vless不需要进行任何特别设置特别是不用设置mux就支持fullcone了？请大佬帮忙解惑，谢谢！

**Comment by wtfr-dot** [comment]
我在#2017里找到这样的描述：顺便先简单说一下 v1.8.1 增强版 XUDP 的 Global ID & UoT Migration 有什么效果：

v1.8.1 以前，你用任何 UoT，假设服务端用 A 端口与多目标通信，若 TCP 断了，比如切换网络，重连后服务端会改用 B 端口。
v1.8.1 开始，你用 VLESS（包括 Mux.Cool），即使 TCP 断了，重连后服务端还是会用 A 端口。

尤其是，对 P2P 有奇效。从某种程度上来说，这才是真正的 FullCone。双端 Xray-core v1.8.1+ 自动启用，无需额外配置。
看来从1.8.1开始不需要任何设置就可以实现fullcone了。

**Comment by RPRX** [comment]
> 我又在#252里看见特性描述里有这样一点：客户端、服务端均为 Xray-core v1.3.0+ 时，VLESS、VMess、Mux 默认均为 FullCone，这是否是说vless不需要进行任何特别设置特别是不用设置mux就支持fullcone了？请大佬帮忙解惑，谢谢！

不止 VLESS，整个 Xray 都是，默认全 FullCone，~~[Xray-core v1.3.0](https://github.com/XTLS/Xray-core/releases/tag/v1.3.0) 放在当年是非常炸裂的~~

> 看来从1.8.1开始不需要任何设置就可以实现fullcone了。

v1.8.1 是 XUDP 默认有了 UoT Migration，它的作用上面描述过了，发现一个评论：https://github.com/XTLS/Xray-core/commit/be23d5d3b741268ef86f27dfcb06389e97447e87#commitcomment-134806788

**Reply by wtfr-dot** [reply]
谢谢大佬解惑！看#1963里面的讨论比较清晰了，对于走cdn来说xray的设置很有限，很多新特性也没法使用，xray默认已经全fullcone，是因为有xudp.go这个单独实现，这是我之前没有注意的地方，所以一直纠结到原生的mux上了，所以对于xudp来说，不开mux或者开mux将concurrency设置-1，udp的出站都是由xray来处理成single xudp

**Comment by wtfr-dot** [comment]
看来需要在防火墙上开放全部端口才能顺利实现fullcone，之前开放部分高位端口国外代理测出只有NAT3，现在国内外测出都是fullcone了。
![QQ截图20240515174849](https://github.com/XTLS/Xray-core/assets/81218740/8cbf41fb-03af-44c6-9758-39589a8b8272)
![QQ截图20240515174923](https://github.com/XTLS/Xray-core/assets/81218740/601cabd3-c215-44ca-bce2-02ff02cd6a51)
![QQ截图20240515174951](https://github.com/XTLS/Xray-core/assets/81218740/e4a8bd86-869c-46f1-b34b-1a80db0c8248)
![QQ截图20240515175014](https://github.com/XTLS/Xray-core/assets/81218740/6d74749d-6a4a-401d-9eb2-c7ee694a5b08)
